In [1]:
import boto3
import numpy as np
import pandas as pd

#### Helper methods to filter and transform dataframes

In [29]:
# df should be a pandas dataframe
# cols_to_keep should be a list of column strings present in df
# assumes the input data frame will have a Status, Fall Term, and County columns
# returns all rows in the dataframe where Status = Admitted and Fall Term = 2018
# filters cols based on the cols_to_keep
# removes duplicates if remove_dups is True (by default)
def filter_dataframe(df, cols_to_keep, year=2016, remove_dups=True):
    # take students who were admitted or applied
    df = df[(df['Status'] == 'Adm') | (df['Status'] == 'App')]
    # only look at the most recent admits
    df = df[df['Fall Term'] ==  year]
    # filter columns
    df = df[cols_to_keep]
    
    if remove_dups:
        # remove duplicates
        df = df.drop_duplicates()
    
    print "Number of rows in the set {0}".format(df.size)
    print "Number of unique counties {0}".format(len(list(df['County'].unique())))
    
    return df

In [30]:
# transform gpas into discrete buckets
# round down to the lower bucket for each range
# buckets are low (1), medium (2), high (3)
# min GPA = 3.24, max GPA = 4.35, stddev = 0.2134
def bucket_gpa(gpa):
    if gpa >= 3.0 and gpa < 3.2:
        return 1

    if gpa >= 3.2 and gpa < 3.4:
        return 1

    if gpa >= 3.4 and gpa < 3.6:
        return 2

    if gpa >= 3.6 and gpa < 3.8:
        return 2

    if gpa >= 3.8 and gpa < 4.0:
        return 2

    if gpa >= 4.0 and gpa < 4.2:
        return 3

    return 3

In [31]:
# transform sat scores into discrete buckets
# low (1), med (2), high (3)
# min score = 440 and max score = 550
def bucket_sat_score(score):
    if score >= 400 and score < 440:
        return 1

    if score >= 440 and score < 460:
        return 1
    
    if score >= 460 and score < 480:
        return 2
    
    if score >= 480 and score < 500:
        return 2
    
    if score >= 500 and score < 520:
        return 2
    
    if score >= 520 and score < 540:
        return 3
    
    return 3

#### Ingest and Format Admit Ethnicity Data
#### Resulting DataFrame should have County and Ethnicity of all 2018 Admitted UC students

In [32]:
# import data by ethnicity
ethnicities = pd.read_csv('../data/HS_ethnicity_by_year_data.csv')
# rename County, Status, and Ethnicity Columns
ethnicities = ethnicities.rename(columns={'County/State/ Territory': 'County', 'Count': 'Status', 'Uad Uc Ethn 6 Cat': 'Ethnicity', 'Pivot Field Values': 'Count'})
# further filter rows
# only keep rows that have a county
ethnicities = ethnicities[pd.notnull(ethnicities['County'])]
ethnicities = filter_dataframe(ethnicities, ['County', 'Ethnicity', 'Status'])

Number of rows in the set 852
Number of unique counties 42


In [33]:
ethnicities.head()

County Ethnicity Status
80     TX     White    Adm
81     TX     Asian    Adm
82     TX     Asian    App
83     TX       All    Adm
84     TX       All    App

In [35]:
# import data by gender
genders = pd.read_csv('../data/FR_GENDER_data.csv')
genders = genders.rename(columns={'Count': 'Status', 'Pivot Field Values': 'Count', 'County/State/ Territory': 'County'})
genders['Fall Term'] = 2016
genders = filter_dataframe(genders, ['County', 'Gender', 'Status'])

Number of rows in the set 999
Number of unique counties 55


In [36]:
genders.head()

County  Gender Status
1  Los Angeles    Male    Adm
2  Los Angeles    Male    App
4  Los Angeles  Female    Adm
5  Los Angeles  Female    App
7  Los Angeles     All    Adm

#### Min GPA for 2018 admits is 3.24
#### Max GPA for 2018 admits is 4.35

In [37]:
# import data by gpa
gpas = pd.read_csv('../data/FR_GPA_by_Inst_data.csv')
# rename columns
gpas = gpas.rename(columns={'Measure Values': 'GPA', 'School Name': 'School', 'Measure Names': 'Status'})
# filter any rows that do not have a GPA or county
gpas = gpas[pd.notnull(gpas['GPA'])]
gpas = gpas[gpas['County'] != 'Not Applicable']
# transform status to match the other datasets
gpas['Status'] = [status.replace('GPA', '').strip() for status in gpas['Status']]
gpas = filter_dataframe(gpas, ['County', 'GPA', 'Campus', 'Status'], remove_dups=False)
# take the average admitted gpa per county and campus
gpas = gpas.groupby(['County', 'Campus', 'Status']).mean().reset_index()
# round all the GPAs into discrete buckets
gpas['GPA'] = [bucket_gpa(gpa) for gpa in gpas['GPA']]
gpas = gpas.drop_duplicates()
print "Number of rows in the set {0}".format(gpas.size)

Number of rows in the set 7596
Number of unique counties 28
Number of rows in the set 1800


In [38]:
gpas.head()

County    Campus Status  GPA
0  Alameda  Berkeley    Adm    3
1  Alameda  Berkeley    App    2
2  Alameda     Davis    Adm    3
3  Alameda     Davis    App    2
4  Alameda    Irvine    Adm    3

#### SAT Score Source data
#### https://data.world/education/california-sat-report-2015-2016/workspace/project-summary?agentid=education&datasetid=california-sat-report-2015-2016

In [76]:
sat_scores = pd.read_csv('../data/sat-report-2015-2016.csv')
# rtype = Record Type: C=County, D=District, S=School, X=State
sat_scores = sat_scores[sat_scores['rtype'] == 'C']
# columns to keep
sat_scores = sat_scores[['AvgScrRead', 'AvgScrMath', 'AvgScrWrit', 'cname']]
# rename for join
sat_scores = sat_scores.rename(columns={'cname': 'County'})
# drop any rows that do not have a school
sat_scores = sat_scores[pd.notnull(sat_scores['County'])]
# convert score cols to ints
sat_scores['AvgScrRead'] = pd.to_numeric(sat_scores['AvgScrRead'], errors=coerce)
sat_scores['AvgScrMath'] = pd.to_numeric(sat_scores['AvgScrMath'], errors=coerce)
sat_scores['AvgScrWrit'] = pd.to_numeric(sat_scores['AvgScrWrit'], errors=coerce)
# drop null cols
sat_scores = sat_scores[pd.notnull(sat_scores['AvgScrRead'])]
sat_scores = sat_scores[pd.notnull(sat_scores['AvgScrMath'])]
sat_scores = sat_scores[pd.notnull(sat_scores['AvgScrWrit'])]
# bucket scores
sat_scores['AvgScrRead'] = [bucket_sat_score(score) for score in sat_scores['AvgScrRead']]
sat_scores['AvgScrMath'] = [bucket_sat_score(score) for score in sat_scores['AvgScrMath']]
sat_scores['AvgScrWrit'] = [bucket_sat_score(score) for score in sat_scores['AvgScrWrit']]
# remove duplicates
sat_scores = sat_scores.drop_duplicates()
sat_scores.head()

AvgScrRead  AvgScrMath  AvgScrWrit     County
1             2           3           2    Alameda
89            3           3           2     Amador
95            2           3           2      Butte
119           2           2           2  Calaveras
126           1           2           1     Colusa

In [77]:
print "Number of rows in the set {0}".format(sat_scores.size)
print "Number of unique counties {0}".format(len(list(sat_scores['County'].unique())))

Number of rows in the set 224
Number of unique counties 56


In [140]:
# join the datasets
result = pd.merge(ethnicities, genders, on=['County', 'Status'])
result = pd.merge(result, gpas, on=['County', 'Status']).drop_duplicates()

# SAT scores do not have admittance data, so assume all applicants from
# the same county have the same SAT scores
result = pd.merge(result, sat_scores, on=['County']).drop_duplicates()
result.drop_duplicates(inplace=True)

In [141]:
result.head()

County         Ethnicity Status Gender       Campus  GPA  \
0  San Bernardino  Domestic Unknown    App   Male     Berkeley    2   
1  San Bernardino  Domestic Unknown    App   Male        Davis    2   
2  San Bernardino  Domestic Unknown    App   Male       Irvine    2   
3  San Bernardino  Domestic Unknown    App   Male  Los Angeles    2   
4  San Bernardino  Domestic Unknown    App   Male       Merced    1   

   AvgScrRead  AvgScrMath  AvgScrWrit  
0           2           2           1  
1           2           2           1  
2           2           2           1  
3           2           2           1  
4           2           2           1

In [97]:
result.size

53460

In [98]:
print ("Total number of admitted students {0}".format(len(result[result['Status'] == 'Adm'])))

Total number of admitted students 2855


In [99]:
print ("Total number of applied students {0}".format(len(result[result['Status'] == 'App'])))

Total number of applied students 3085


#### Calculate a percentage of admitted to applied students

In [135]:
# group result by county and campus
grouping_cols = ['County', 'Campus', 'Gender']

# group all the admitted students by the above and count how many students were admitted with those stats
num_admitted = result[result['Status'] == 'Adm'] \
    .groupby(grouping_cols) \
    .count() \
    .reset_index() \
    .rename(columns={'Status': 'NumAdmittedStudents'}) \
    .drop(['Ethnicity', 'GPA', 'AvgScrRead', 'AvgScrMath', 'AvgScrWrit'], axis=1)
    
# group all the applied students by the above and count how many students applied with those stats
num_applied = result[result['Status'] == 'App'] \
    .groupby(grouping_cols) \
    .count() \
    .reset_index() \
    .rename(columns={'Status': 'NumAppliedStudents'}) \
    .drop(['Ethnicity', 'GPA', 'AvgScrRead', 'AvgScrMath', 'AvgScrWrit'], axis=1)
    
# merge the datasets back together on the original grouping columns
probas = pd.merge(num_admitted, num_applied, on=grouping_cols)
# calculate the admittance rate by dividing admitted / applied students
probas['AdmittanceRate'] = probas['NumAdmittedStudents'] / probas['NumAppliedStudents']
# some kind of invalid data? drop any cols that have more admitted than applied students
probas = probas.drop(probas[probas['NumAdmittedStudents'] > probas['NumAppliedStudents']].index)
# drop the intermediate columns
probas = probas.drop(['NumAdmittedStudents', 'NumAppliedStudents'], axis=1)
probas.head()

County    Campus   Gender  AdmittanceRate
0  Alameda  Berkeley      All             1.0
1  Alameda  Berkeley   Female             1.0
2  Alameda  Berkeley     Male             1.0
3  Alameda  Berkeley  Unknown             1.0
4  Alameda     Davis      All             1.0

#### Grouping by any additional attributes past county, UC campus, and gender
#### results in categories that are too specific, and therefore probabilities that are all 100%
#### even after bucketing SAT and GPA scores
#### Grouping by campus, county, and gender results in admittance rates between 66.67% and 100%
#### with an average admittance rate of 95.8% and standard deviation of 0.0957

In [137]:
probas.describe()

AdmittanceRate
count      516.000000
mean         0.958527
std          0.095793
min          0.666667
25%          1.000000
50%          1.000000
75%          1.000000
max          1.000000

#### Join the probabilities dataframe back with the original dataframe to get a complete training set

In [138]:
result = pd.merge(result, probas, on=grouping_cols)
result.head()

County         Ethnicity Status Gender    Campus  GPA  AvgScrRead  \
0  San Bernardino  Domestic Unknown    App   Male  Berkeley    2           2   
1  San Bernardino  Hispanic/ Latino    App   Male  Berkeley    2           2   
2  San Bernardino               All    App   Male  Berkeley    2           2   
3  San Bernardino             White    App   Male  Berkeley    2           2   
4  San Bernardino             Asian    App   Male  Berkeley    2           2   

   AvgScrMath  AvgScrWrit  AdmittanceRate  
0           2           1             1.0  
1           2           1             1.0  
2           2           1             1.0  
3           2           1             1.0  
4           2           1             1.0

In [139]:
result.describe()

GPA   AvgScrRead   AvgScrMath   AvgScrWrit  AdmittanceRate
count  5057.000000  5057.000000  5057.000000  5057.000000     5057.000000
mean      2.316591     2.138422     2.285940     1.925648        0.969824
std       0.485579     0.396560     0.492132     0.587994        0.080276
min       1.000000     1.000000     1.000000     1.000000        0.666667
25%       2.000000     2.000000     2.000000     2.000000        1.000000
50%       2.000000     2.000000     2.000000     2.000000        1.000000
75%       3.000000     2.000000     3.000000     2.000000        1.000000
max       3.000000     3.000000     3.000000     3.000000        1.000000

#### Resulting dataset has 5,057 rows
#### With county, ethnicity, applicant status, gender, UC campus, GPA, SAT scores, and calculated admittance rate

In [142]:
result.to_csv('../data/applicant_admissions_data.csv')

In [143]:
S3 = boto3.client('s3', region_name='eu-central-1')
S3.upload_file('../data/applicant_admissions_data.csv', 'gosat-data', 'applicant_admissions_data.csv')